packages/react-core/src/components/TreeView/TreeViewListItem.tsx

```diff
@@ -186,20 +198,22 @@ const TreeViewListItemBase: React.FunctionComponent<TreeViewListItemProps> = ({
               onClick={(evt: React.MouseEvent) => {
                 if (!hasCheck) {
                   onSelect && onSelect(evt, itemData, parentItem);
-                  if (children && evt.isDefaultPrevented() !== true) {
+                  if (!hasSelectableNodes && children && evt.isDefaultPrevented() !== true) {
                     setIsExpanded(!internalIsExpanded);
                   }
                 }
               }}
-              tabIndex={-1}
+              tabIndex={hasSelectableNodes ? 1 : -1}
+              {...(hasCheck && { htmlFor: randomId })}
+              {...((hasCheck || (hasSelectableNodes && children)) && { id: `label-${randomId}` })}
```

Comment:
The ID that gets set here includes "checkbox" in the value regardless of whether there's a checkbox or not. May be worth updating the randomId prefix to render "checkbox" or "selectable" or something depending on whether `hasSelectableNodes` is passed in.